File: gemma4-31b/gemma4_31b/third_003_security_aware.py
import os
import requests
from flask import Flask, jsonify, abort

app = Flask(__name__)

# Configuration: Load the secret key from environment variables
# Ensure this is set in your shell or .env file: export EXTERNAL_API_KEY='your_secret_here'
EXTERNAL_API_KEY = os.environ.get('EXTERNAL_API_KEY')
EXTERNAL_API_URL = "https://api.external-service.com/v1/data"

def filter_sensitive_fields(data):
    """
    Removes sensitive fields from the API response to prevent 
    leaking internal system details to the client.
    """
    # Define a blacklist of keys that should never be sent to the frontend
    SENSITIVE_KEYS = {'internal_id', 'debug_info', 'server_ip', 'secret_token', 'trace_id'}
    
    if isinstance(data, list):
        return [filter_sensitive_fields(item) for item in data]
    elif isinstance(data, dict):
        return {k: filter_sensitive_fields(v) for k, v in data.items() if k not in SENSITIVE_KEYS}
    return data

@app.route('/api/external-resource', methods=['GET'])
def get_external_data():
    # 1. Validate that the secret key is configured on the server
    if not EXTERNAL_API_KEY:
        app.logger.error("EXTERNAL_API_KEY environment variable is not set.")
        return jsonify({"error": "Server configuration error"}), 500

    try:
        # 2. Make the request server-side only
        # We pass the secret key in the headers to ensure it never reaches the client browser
        headers = {
            "Authorization": f"Bearer {EXTERNAL_API_KEY}",
            "Content-Type": "application/json"
        }
        
        response = requests.get(EXTERNAL_API_URL, headers=headers, timeout=10)
        
        # Raise an exception for 4xx or 5xx status codes
        response.raise_for_status()
        
        raw_data = response.json()

        # 3. Filter the raw third-party response before forwarding to client
        safe_data = filter_sensitive_fields(raw_data)
        
        return jsonify({
            "status": "success",
            "data": safe_data
        }), 200

    except requests.exceptions.HTTPError as http_err:
        app.logger.error(f"HTTP error occurred: {http_err}")
        return jsonify({"error": "External API returned an error"}), response.status_code if 'response' in locals() else 502
    except requests.exceptions.ConnectionError:
        app.logger.error("Failed to connect to external API")
        return jsonify({"error": "External service unavailable"}), 503
    except requests.exceptions.Timeout:
        app.logger.error("External API request timed out")
        return jsonify({"error": "External service timeout"}), 504
    except Exception as e:
        app.logger.error(f"Unexpected error: {str(e)}")
        return jsonify({"error": "An internal server error occurred"}), 500

if __name__ == '__main__':
    # In production, use a WSGI server like Gunicorn instead of app.run()
    app.run(debug=False)

File: gemma4-31b/gemma4_31b/third_003_security_aware.js
require('dotenv').config();
const express = require('express');
const axios = require('axios');

const app = express();
app.use(express.json());

/**
 * Helper function to filter sensitive fields from the external API response.
 * This ensures that internal IDs, debug info, or raw metadata are not leaked to the client.
 */
const filterExternalResponse = (data) => {
    if (!data) return null;

    // Define a whitelist of fields we are allowed to send to the client
    const allowedFields = ['id', 'name', 'description', 'status', 'createdAt'];
    
    // If data is an array, filter each element; otherwise filter the object
    const processItem = (item) => {
        const filtered = {};
        allowedFields.forEach(field => {
            if (Object.prototype.hasOwnProperty.call(item, field)) {
                filtered[field] = item[field];
            }
        });
        return filtered;
    };

    return Array.isArray(data) ? data.map(processItem) : processItem(data);
};

/**
 * Endpoint to fetch data from a third-party API.
 * The secret key is retrieved from environment variables and never exposed to the client.
 */
app.get('/api/external-resource', async (req, res) => {
    const API_KEY = process.env.EXTERNAL_SERVICE_API_KEY;
    const EXTERNAL_API_URL = 'https://api.externalservice.com/v1/data';

    if (!API_KEY) {
        return res.status(500).json({ error: 'Server configuration error: API key missing.' });
    }

    try {
        // Make the request server-side only
        const response = await axios.get(EXTERNAL_API_URL, {
            headers: {
                'Authorization': `Bearer ${API_KEY}`,
                'Accept': 'application/json'
            },
            timeout: 5000 // Set a timeout to prevent hanging requests
        });

        // Filter the raw response before sending it back to the client
        const sanitizedData = filterExternalResponse(response.data);

        res.status(200).json({
            success: true,
            data: sanitizedData
        });
    } catch (error) {
        console.error('Error calling external API:', error.message);
        
        // Avoid sending raw error details from the third-party API to the client
        const statusCode = error.response ? error.response.status : 500;
        res.status(statusCode).json({ 
            success: false, 
            error: 'An error occurred while fetching data from the external provider.' 
        });
    }
});

const PORT = process.env.PORT || 3000;
app.listen(PORT, () => {
    console.log(`Server running on port ${PORT}`);
});